perf/rwlocks.c

```diff
@@ -0,0 +1,179 @@
+    if (writers == 0)
+        writer_end = ossl_time_now();
+    CRYPTO_THREAD_unlock(lock); 
+}
+
+void do_rw_rlock(size_t num)
+{
+    int i;
+    unsigned long last_val = 0;
+    int local_read_lock_calls = 0;
+
+    for (i = 0; i < NUM_CALLS_PER_THREAD; i++) {
+        CRYPTO_THREAD_read_lock(lock);
+        if (dataval != NULL) {
+            if (last_val != 0 && last_val > *dataval)
+                printf("dataval went backwards! %lu:%lu\n", last_val, *dataval);
+            last_val = *dataval;
+        }
+        CRYPTO_THREAD_unlock(lock);
+        local_read_lock_calls += 2; /* lock and unlock */
+    }
+
+    CRYPTO_THREAD_write_lock(lock);
+    read_lock_calls += local_read_lock_calls;
+    readers--;
+    if (readers == 0)
+        reader_end = ossl_time_now();
+    CRYPTO_THREAD_unlock(lock);
+}
+
+void do_rwlocks(size_t num)
+{
+    if (num >= threadcount - writers)
+        do_rw_wlock(num);
+    else
+        do_rw_rlock(num);
+}
+
+int main(int argc, char *argv[])
+{
+    OSSL_TIME duration;
+    OSSL_TIME start;
+    uint64_t us;
+    double avwcalltime;
+    double avrcalltime;
+    int terse = 0;
+    int argnext;
+    char *writeenv;
+
+    if ((argc != 2 && argc != 3)
+       || (argc == 3 && strcmp("--terse", argv[1]) != 0)) {
+        printf("Usage: rwlocks [--terse] threadcount\n");
+        return EXIT_FAILURE;
+    }
+
+    if (argc == 3) {
+        terse = 1;
+        argnext = 2;
+    } else {
+        argnext = 1;
+    }
+
+    threadcount = atoi(argv[argnext]);
+    if (threadcount < 1) {
+        printf("threadcount must be > 0\n");
+        return EXIT_FAILURE;
+    }
+
+    writeenv=getenv("LOCK_WRITERS");
+    if (writeenv == NULL) {
+        writers = threadcount / 2;
+    } else {
+        writers=atoi(writeenv);
+        if (writers == 0)
+            writers = threadcount / 2;
+    }
+
+    lock = CRYPTO_THREAD_lock_new();
+    if (lock == NULL) {
+        printf("unable to allocate lock\n");
+        return EXIT_FAILURE;
+    }
+
+    readers = threadcount - writers;
+
+    if (!terse)
+        printf("Running rwlock test with %d writers and %d readers\n", writers, readers);
+
+    start = ossl_time_now(); 
+
+    if (!perflib_run_multi_thread_test(do_rwlocks, threadcount, &duration)) {
+        printf("Failed to run the test\n");
+        return EXIT_FAILURE;
+    }
+
+    if (err) {
+        printf("Error during test\n");
+        return EXIT_FAILURE;
+    }
+
+    us = ossl_time2us(ossl_time_subtract(writer_end, start));
+    avwcalltime = (double)us / (double)write_lock_calls;
+
+    if (!terse)
+        printf("total write lock/unlock calls %d in %lf us\n", write_lock_calls, (double)us);
+
+    us = ossl_time2us(ossl_time_subtract(reader_end, start));
+    avrcalltime = (double)us / (double)read_lock_calls;
+    if (!terse)
+        printf("total read lock/unlock calls %d %lf us\n", read_lock_calls, (double)us);
+
+    if (terse)
+        printf("%lf %lf\n", avwcalltime, avrcalltime);
```

Comment:
Enclose this statement in `{}`. OpenSSL coding style says that if one branch of an `if...else...` uses `{}` then both branches should do.